src/contracts/core/AVSDirectory.sol

```diff
@@ -52,67 +52,295 @@ contract AVSDirectory is
+     *
+     * @dev msg.sender must be the AVS.
+     * @dev The AVS may create operator sets before it becomes an operator set AVS.
+     */
+    function createOperatorSets(uint32[] calldata operatorSetIds) external {
+        for (uint256 i = 0; i < operatorSetIds.length; ++i) {
+            require(
+                !isOperatorSet[msg.sender][operatorSetIds[i]],
+                "AVSDirectory.createOperatorSet: operator set already exists"
+            );
+            isOperatorSet[msg.sender][operatorSetIds[i]] = true;
+            emit OperatorSetCreated(OperatorSet({avs: msg.sender, operatorSetId: operatorSetIds[i]}));
+        }
+    }
+
+    /**
+     * @notice Sets the AVS as an operator set AVS, preventing legacy M2 operator registrations.
+     *
+     * @dev msg.sender must be the AVS.
+     */
+    function becomeOperatorSetAVS() external {
+        require(!isOperatorSetAVS[msg.sender], "AVSDirectory.becomeOperatorSetAVS: already an operator set AVS");
+        isOperatorSetAVS[msg.sender] = true;
+        emit AVSMigratedToOperatorSets(msg.sender);
+    }
+
+    /**
+     * @notice Called by an AVS to migrate operators that have a legacy M2 registration to operator sets.
+     *
+     * @param operators The list of operators to migrate
+     * @param operatorSetIds The list of operatorSets to migrate the operators to
+     *
+     * @dev The msg.sender used is the AVS
+     * @dev The operator can only be migrated at most once per AVS
+     * @dev The AVS can no longer register operators via the legacy M2 registration path once it begins migration
+     * @dev The operator is deregistered from the M2 legacy AVS once migrated
+     */
+    function migrateOperatorsToOperatorSets(
+        address[] calldata operators,
+        uint32[][] calldata operatorSetIds
+    ) external override onlyWhenNotPaused(PAUSED_OPERATOR_REGISTER_DEREGISTER_TO_AVS) {
+        // Assert that the AVS is an operator set AVS.
+        require(
+            isOperatorSetAVS[msg.sender], "AVSDirectory.migrateOperatorsToOperatorSets: AVS is not an operator set AVS"
+        );
+
+        for (uint256 i = 0; i < operators.length; i++) {
+            // Assert that the operator is registered & has not been migrated.
+            require(
+                avsOperatorStatus[msg.sender][operators[i]] == OperatorAVSRegistrationStatus.REGISTERED,
+                "AVSDirectory.migrateOperatorsToOperatorSets: operator already migrated or not a legacy registered operator"
+            );
+
+            // Migrate operator to operator sets.
+            _registerToOperatorSets(msg.sender, operators[i], operatorSetIds[i]);
+
+            // Deregister operator from AVS - this prevents the operator from being migrated again since
+            // the AVS can no longer use the legacy M2 registration path
+            avsOperatorStatus[msg.sender][operators[i]] = OperatorAVSRegistrationStatus.UNREGISTERED;
```

Comment:
Maybe should note this storage slot is to be deprecated in a future release once AVSs have migrated

Reply:
Added comment